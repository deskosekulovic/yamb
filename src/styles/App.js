import styled from 'styled-components';

export const theme = {
    backgroundColor: '#6495ED',
    selectColor: 'blue',
    gameBackground: '#DCDCDC',
    fontSize: '14px',
    primaryColor: 'black',
    rowsColor: '#1E90FF',
    color: 'white',
    height: '30px',
    border: '1px solid black'
};

export const StyledApp = styled.div`
  margin: 0 auto;
  width: 900px;
`;
export const Span = styled.span`
  text-decoration: underline;
  &:hover {
  cursor: pointer;
  }
`;

export const ColumnName = styled.div`
  font-weight: bold;
  margin-top: 20px;
`;
export const Game = styled.div`
  float: left;
  padding: 30px;
  background-color: ${(props) => props.theme.gameBackground};
`;

export const RowName = styled.button`
  border: ${(props) => props.theme.border};
  width: 60px;
  height: ${(props) => props.theme.height};
  text-align: center;
  font-weight: bold;
  color: ${(props) => props.theme.color};
  background-color: ${(props) => props.theme.rowsColor};
`;

export const StyledField = styled.button`
  border: ${(props) => props.theme.border};
  width: 60px;
  height: 30px;
  text-align: center;
  color: ${(props) => props.theme.color};
  background-color: ${(props) => props.permission ? props.theme.selectColor : props.theme.backgroundColor};
`;

export const Rezultati = styled.div`
  padding: 30px;
  background-color: ${(props) => props.theme.gameBackground};
  height: 700px
`;
