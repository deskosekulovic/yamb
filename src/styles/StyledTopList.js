import styled from 'styled-components';

const Table = styled.table`
  border-collapse: collapse;
  th, td {
    border: ${(props) => props.theme.border};
    border-collapse: collapse;
    padding: 10px;
    background-color: ${(props) => props.theme.color};
  }
`;

export default Table;
