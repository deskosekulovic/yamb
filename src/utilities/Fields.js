export const rows = ['1','2','3','4','5','6','suma1','max','min','suma2',
    'kenta','triling','full','poker','yumb','suma3'];

export const columns = ['dole','slobodno','gore','najava','rucno'];

export const rowsToSelect = ['1','2','3','4','5','6','max','min',
    'kenta','triling','full','poker','yumb'];

export const dices = ['dice-1','dice-2','dice-3','dice-4','dice-5','dice-6'];
