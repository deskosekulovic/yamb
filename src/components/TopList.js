import React, { Component } from 'react';
import PropTypes from 'prop-types';
import StyledTopList from '../styles/StyledTopList';
import { Link } from 'react-router-dom';
import TopListSettings from './TopListSettings';
import DisplayTopList from './DisplayTopList';
import Button from '../styles/Button';
import { saveTopListSettings, getDataSettings } from '../utilities/store';

class TopList extends Component{
    constructor(props){
        super(props);
        const dataSettings = getDataSettings('topListSettings');
        this.state={
            localNumberOfDice: dataSettings['localNumberOfDice'] || '6',
            localNumberOfColumns: dataSettings['localNumberOfColumns'] || '7',
            numberOfResults: dataSettings['numberOfResults'] || '10',
            showTopListSettings: false
        };
        this.handleChange=this.handleChange.bind(this);
        this.toggleTopListSettings=this.toggleTopListSettings.bind(this);
    }

    handleChange(e) {
        const target = e.target;
        const name = target.name;
        const value = target.value;
        saveTopListSettings(name, value);
        this.setState({
            [name]: value
        });
    }
    toggleTopListSettings() {
        this.setState({
            showTopListSettings: !this.state.showTopListSettings
        });
    }
    render(){
        const { localNumberOfDice, localNumberOfColumns, numberOfResults, showTopListSettings } = this.state;

        return(
            <StyledTopList>
                <TopListSettings
                    showTopListSettings={showTopListSettings}
                    localNumberOfDice={localNumberOfDice}
                    localNumberOfColumns={localNumberOfColumns}
                    numberOfResults={numberOfResults}
                    toggleTopListSettings={this.toggleTopListSettings}
                    handleChange={this.handleChange}
                />
                <DisplayTopList
                    localNumberOfDice={localNumberOfDice}
                    localNumberOfColumns={localNumberOfColumns}
                    numberOfResults={numberOfResults}
                />
                <Link to='/'><Button>Nazad</Button></Link>
            </StyledTopList>
        );
    }
}

TopList.propTypes={
    totalResult: PropTypes.number,
    numberOfDice: PropTypes.string,
    numberOfColumns: PropTypes.number,
    match: PropTypes.object
};

export default TopList;
