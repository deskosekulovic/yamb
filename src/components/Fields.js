import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { RowName, ColumnName, StyledField } from '../styles/App';
import Field from './Field';
import { rows, columns } from '../utilities/Fields';

class Fields extends Component {
    render(){
        const { handleInput, fields, permission } = this.props;
        return(
            <div>
                <ColumnName>
                    <StyledField disabled={true}>&nbsp;</StyledField>
                    <StyledField disabled={true}>&darr;</StyledField>
                    <StyledField disabled={true}>&darr; &uarr;</StyledField>
                    <StyledField disabled={true}>&uarr;</StyledField>
                    <StyledField disabled={true}>N</StyledField>
                    <StyledField disabled={true}>R</StyledField>
                    <StyledField disabled={true}>&harr;</StyledField>
                    <StyledField disabled={true}>&#8597;</StyledField>
                </ColumnName>
                {rows.map(row => {
                    return(
                        <div key={row} id={row} onClick={handleInput}>
                            <RowName disabled={true}>{row}</RowName>
                            {columns.map(column =>
                                <Field
                                    key={column}
                                    id={`${column}-${row}`}
                                    value={fields}
                                    permission={permission}
                                />
                            )}
                        </div>
                    );
                })}
            </div>
        );


    }
}

Fields.propTypes = {
    handleInput: PropTypes.func,
    fields: PropTypes.object,
    permission: PropTypes.object
};

export default Fields;
