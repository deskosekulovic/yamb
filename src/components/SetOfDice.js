import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Dice from './Dice';
import { setOfDice } from '../utilities/Fields';
import StyledSetOfDice from '../styles/StyledSetOfDice';

class SetOfDices extends Component{
    render(){
        const { dice, toggleSelectDice, rollCounter, selected, numberOfDice } = this.props;
        return(
            <StyledSetOfDice>
                {setOfDice.slice(0,numberOfDice).map((el,i) =>
                    <Dice
                        key={el}
                        name={el}
                        rollCounter={rollCounter}
                        toggleSelectDice={toggleSelectDice}
                        selected={selected[i]}
                        dice={dice[i]}
                    />)}
            </StyledSetOfDice>
        );
    }
}
SetOfDices.propTypes = {
    dice: PropTypes.array,
    toggleSelectDice: PropTypes.func,
    rollCounter: PropTypes.number,
    selected: PropTypes.array,
    numberOfDice: PropTypes.string
};

export default SetOfDices;
