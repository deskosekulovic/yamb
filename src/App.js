import React, { Component } from 'react';
import { Switch, Route } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { theme } from './styles/App';
import Main from './components/Main';
import Settings from './components/Settings';
import TopList from './components/TopList';
import { setExtraColumns } from './utilities/Functions';
import { getDataSettings } from './utilities/store';
import { columns, rowsToSelect } from './utilities/Fields';

class App extends Component {
    constructor(props){
        super(props);
        let dataSettings = getDataSettings();
        this.state = {
            numberOfDice: dataSettings['numberOfDice'] || '6'
        };
        this.handleChange=this.handleChange.bind(this);
    }
    componentDidMount(){
        let dataSettings = getDataSettings();
        let rest = dataSettings['columnsToAdd'];
        this.setState({...rest});
        setExtraColumns({...rest});
    }
    handleChange(e){
        const target = e.target;
        const value = target.type === 'checkbox' ? target.checked : target.value;
        const name = target.name;
        this.setState({
            [name]: value,
        });
    }
    render() {
        const { numberOfDice, ...rest } = this.state;
        let bonusLength = 0;
        Object.keys(rest).map(item=>{
            rest[item] && bonusLength++;
        });
        const numberOfFields = (columns.length+bonusLength)*rowsToSelect.length;
        return (
            <ThemeProvider theme={theme}>
                <Switch>
                    <Route exact path="/" render={props=><Main {...props} columnsToAdd={rest} numberOfColumns={columns.length+bonusLength} numberOfDice={numberOfDice} numberOfFields={numberOfFields} />} />
                    <Route path="/settings" render={props=><Settings {...props} columnsToAdd={rest} numberOfDice={numberOfDice} handleChange={this.handleChange} />} />
                    <Route path="/toplists" render={props=><TopList {...props} numberOfColumns={columns.length+bonusLength} columnsToAdd={rest} numberOfDice={numberOfDice} />} />
                </Switch>
            </ThemeProvider>
        );
    }
}

export default App;
