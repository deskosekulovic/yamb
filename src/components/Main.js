import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { StyledApp, Game, Span, Results, Navigation } from '../styles/App';
import calculate, { permissionHandler, columnResult, totalResult, selectedDice } from '../utilities/Functions';
import Fields from './Fields';
import SetOfDice from './SetOfDice';
import Button from './Button';
import { saveData } from '../utilities/store';

class Main extends Component {
    constructor(props){
        super(props);
        this.state = {
            dice: [],
            selected: [],
            rollCounter: 0,
            fields: {},
            permission: {},
            najavljeno: false,
            inputCount: 0
        };
        this.rollDice = this.rollDice.bind(this);
        this.toggleSelectDice = this.toggleSelectDice.bind(this);
        this.handleInput=this.handleInput.bind(this);
        this.resetGame=this.resetGame.bind(this);
        this.handleMouseOver=this.handleMouseOver.bind(this);
        this.handleMouseOut=this.handleMouseOut.bind(this);
    }
    componentDidMount() {
        this.resetGame();
        window.scrollTo(0, 0);
    }
    componentDidUpdate(prevProps, prevState) {
        const { rollCounter, najavljeno, inputCount } = this.state;
        const { columnsToAdd, numberOfFields, numberOfDice, numberOfColumns } = this.props;
        let per=permissionHandler(null, rollCounter);
        if((prevState.rollCounter!==rollCounter && rollCounter) || (prevState.najavljeno!==najavljeno && rollCounter)){
            this.najavaAndRucno = 0;
            Object.keys(per).map(item=>{
                (columnsToAdd['rucno'] && item.split('-')[0]=='rucno' || item.split('-')[0]=='najava') && per[item] && this.najavaAndRucno++;
            });
            if(!najavljeno){
                this.setState({permission: {...per}});
            }
        }
        if(inputCount===numberOfFields) {
            let name = prompt('Unesite ime', '');
            name = name && name.trim().length > 0 ? name : 'Neznani junak';
            saveData(name, totalResult(), numberOfDice, numberOfColumns);
        }
    }
    rollDice(){
        const { selected, dice, rollCounter, inputCount, najavljeno } = this.state;
        const { numberOfDice, numberOfFields } = this.props;
        if((numberOfFields-inputCount)!==this.najavaAndRucno || !rollCounter || najavljeno){
            let diceValue=[];
            for(let i=0; i<numberOfDice; i++){
                diceValue[i] = selected[i] ? dice[i] : Math.floor(Math.random()*6+1);
            }
            this.setState({
                dice: diceValue,
                rollCounter: rollCounter + 1
            });
        }
        if((numberOfFields-inputCount)==this.najavaAndRucno && rollCounter && !najavljeno){
            alert('Morate najaviti ili uneti iznos!');
        }
    }
    toggleSelectDice(e){
        let index=e.target.name.split('-')[1]-1;
        this.setState({
            selected: this.state.selected.map((el,i)=>{
                return i===index ? !el : el;
            })
        });
    }
    handleInput(e){
        const { dice, rollCounter, najavljeno, fields, inputCount } = this.state;
        const { numberOfDice } = this.props;
        let field = e.target.id.split('-')[1];
        let column = e.target.id.split('-')[0];
        let input, res={};
        if(((column==='rucno' || (column==='najava' && !najavljeno)) && rollCounter!==1) || (column==='najava' && rollCounter===1 && field!=='kenta')){
            input=0;
        }else{
            input=calculate(field,dice,rollCounter);
            res=columnResult(e.target.id,input);
        }
        (!(column==='najava' && rollCounter===1) || (column==='najava' && rollCounter===1 && field==='kenta' && input)) && permissionHandler(e.target.id);

        column==='najava' && rollCounter===1 && this.setState({
            permission: {[e.target.id]: true},
            najavljeno: !najavljeno
        });
        (column==='najava' && (rollCounter===1 && field==='kenta' && input || rollCounter!==1) || column!=='najava') && this.setState({
            fields: {...fields, [e.target.id]: input,...res},
            rollCounter: 0,
            dice: [],
            selected: selectedDice(numberOfDice),
            permission: {},
            inputCount: inputCount+1,
            najavljeno: false
        });
    }
    handleMouseOver(e){
        const { dice, rollCounter, fields } = this.state;
        let field = e.target.id.split('-')[1];
        let input=calculate(field,dice,rollCounter);

        this.setState({
            fields: {...fields, [e.target.id]: input}
        });
    }
    handleMouseOut(e){
        const { fields } = this.state;
        this.setState({
            fields: {...fields, [e.target.id]: ''}
        });
    }
    resetGame(){
        const { numberOfDice } = this.props;
        this.setState({
            dice: [],
            selected: selectedDice(numberOfDice),
            rollCounter: 0,
            fields: {},
            permission: {},
            najavljeno: false,
            inputCount: 0
        });
        permissionHandler(null, null, null, true);
        columnResult(null, null, true);
    }
    render() {
        const { dice, rollCounter, selected, fields, permission, inputCount, najavljeno } = this.state;
        const { numberOfDice, columnsToAdd, numberOfFields, columns } = this.props;

        let extraColumns = Object.keys(columnsToAdd);
        let mergedColumns = [...columns];
        extraColumns.map(
            item=>{
                if(columnsToAdd[item]){
                    mergedColumns=[...mergedColumns,...[{[item]: item=='rucno' ?' R' : item}]];
                }
            }
        );
        return (
            <StyledApp>
                <Game>
                    <Navigation>
                        <Span onClick={this.resetGame}><b>Nova igra</b></Span>
                        <Link to='/settings' style={{'color':'black','paddingLeft':'10px'}}><b>Pode≈°avanja</b></Link>
                        <Link to='/toplists' style={{'color':'black','paddingLeft':'10px'}}><b>Top Liste</b></Link>
                    </Navigation>
                    <Fields
                        fields={fields}
                        columns={mergedColumns}
                        handleInput={this.handleInput}
                        handleMouseOver={this.handleMouseOver}
                        handleMouseOut={this.handleMouseOut}
                        permission={permission}
                        rollCounter={rollCounter}
                        najavljeno={najavljeno}
                    />
                    <SetOfDice
                        toggleSelectDice={this.toggleSelectDice}
                        selected={selected}
                        rollCounter={rollCounter}
                        dice={dice}
                        numberOfDice={numberOfDice}
                    />
                    <Button
                        rollDice={this.rollDice}
                        rollCounter={rollCounter}
                        disabled={numberOfFields===inputCount}
                    />
                </Game>
                <Results>
                    {totalResult()!==0 && <h1>Rezultat: {totalResult()}</h1>}
                </Results>
            </StyledApp>
        );
    }
}

Main.propTypes={
    numberOfDice: PropTypes.string,
    columnsToAdd: PropTypes.object,
    columns: PropTypes.array,
    numberOfFields: PropTypes.number,
    numberOfColumns: PropTypes.number,
    match: PropTypes.object
};

export default Main;
