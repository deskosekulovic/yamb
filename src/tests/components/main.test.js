import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import 'jest-styled-components';
import App from '../../App';
import Main from '../../components/Main';
import Fields from '../../components/Fields';
import SetOfDice from '../../components/SetOfDice';
import TopList from '../../components/TopList';
import { StyledApp, Game, Rezultati, theme } from '../../styles/App';

describe('Main', () => {
    it('should match snapshot', () => {
        const wrapper = shallow(<Main inputCount={50} numberOfFields={50} />);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.containsMatchingElement(<Fields />)).toBe(true);
        expect(wrapper.containsMatchingElement(<SetOfDice />)).toBe(true);
        // if(this.props.inputCount===this.props.numberOfFields) expect(wrapper.containsMatchingElement(<TopList />)).toBe(true);
    });
    describe('<Game />', () => {
        const wrapper = shallow(<Game theme={theme} />);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper).toHaveStyleRule('padding', '5px 30px');
    });
    describe('<Rezultati />', () => {
        const wrapper = shallow(<Rezultati theme={theme} />);
        expect(wrapper).toMatchSnapshot();
    });
});
