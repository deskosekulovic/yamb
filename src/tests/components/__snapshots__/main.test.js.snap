// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
.c0 {
  float: left;
  padding: 5px 30px;
  background-color: #DCDCDC;
}

<div
  className="c0"
/>
`;

exports[` 2`] = `
.c0 {
  padding: 20px;
  background-color: #DCDCDC;
  height: 630px;
}

<div
  className="c0"
/>
`;

exports[`Main should match snapshot 1`] = `
<styled.div>
  <styled.div>
    <styled.span
      onClick={[Function]}
    >
      <b>
        Nova igra
      </b>
    </styled.span>
    <Link
      replace={false}
      style={
        Object {
          "color": "black",
          "paddingLeft": "10px",
        }
      }
      to="/settings"
    >
      <b>
        Pode≈°avanja
      </b>
    </Link>
    <Link
      replace={false}
      style={
        Object {
          "color": "black",
          "paddingLeft": "10px",
        }
      }
      to="/toplists"
    >
      <b>
        Top Liste
      </b>
    </Link>
    <Fields
      fields={Object {}}
      handleInput={[Function]}
      handleMouseOut={[Function]}
      handleMouseOver={[Function]}
      najavljeno={false}
      permission={Object {}}
      rollCounter={0}
    />
    <SetOfDices
      dice={Array []}
      rollCounter={0}
      rollDice={[Function]}
      selected={Array []}
      toggleSelectDice={[Function]}
    />
  </styled.div>
  <styled.div />
</styled.div>
`;
